platforms/documentation/docs/src/docs/userguide/best-practices/bp_tasks.adoc

```diff
@@ -121,3 +121,66 @@ include::sample[dir="snippets/bestPractices/useCacheabilityAnnotations/groovy",f
 
 === Tags
 `#tasks`, `#caching`
+
+== Don't resolve Configurations before Task Execution
+
+Resolving configuration before task execution phase can lead to incorrect results and slows gradle build performance.
+
+Once configuration is resolved it can't be resolved again, and it can lose some information when it is resolved too early.
+Calling manual resolution via link:{javadocPath}/org/gradle/api/artifacts/Configuration.html#resolve()[resolve] method returns a set of Files that doesn't preserve references to the tasks that produced these files.
+This will break automatic task dependencies wiring through task input properties.
+
+=== Example
+
+==== Don't Do This
+
+====
+include::sample[dir="snippets/bestPractices/avoidResolvingConfigurationsManually/kotlin",files="build.gradle.kts[tags=resolve-call]"]
+include::sample[dir="snippets/bestPractices/avoidResolvingConfigurationsManually/groovy",files="build.gradle[tags=resolve-call]"]
+====
+
+<1> *Explicit resolve call*: `runtimeClasspath` configuration will be resolved as soon as task created. It will lose all task dependencies as well.
+
+==== Do This Instead
+
+Use lazy link:{javadocPath}/org/gradle/api/file/FileCollection.html[FileCollection] API instead.
+
+====
+include::sample[dir="snippets/bestPractices/avoidResolvingConfigurationsManually/kotlin",files="build.gradle.kts[tags=good-classpath-printer]"]
+include::sample[dir="snippets/bestPractices/avoidResolvingConfigurationsManually/groovy",files="build.gradle[tags=good-classpath-printer]"]
+====
+
+<1> *Pass runtimeClasspath as input to the File Collection*: the File Collection will not resolve `runtimeClasspath` immediately, and will preserve its task dependencies.
+<2> *Access resolved files*: It is safe to access files from the File Collection during task execution.
+
+=== Tags
+`#tasks`, `#inputs-and-outputs`, `#configurations`
+
+== Avoid using eager API on Gradle File Collections before task execution
+
+Gradle link:{javadocPath}/org/gradle/api/artifacts/Configuration.html[Configuration] and link:{javadocPath}/org/gradle/api/artifacts/FileCollection.html[FileCollection]
+extend JDK's `Collection<File>` interface, however calls to the JDK Collection API on configuration and file collection instances will trigger their implicit resolution.
+
+=== Example
+==== Don't Do This
+
+====
+include::sample[dir="snippets/bestPractices/avoidJdkCollectionApiOnFileCollections/kotlin",files="build.gradle.kts[tags=bad-zipping-task]"]
+include::sample[dir="snippets/bestPractices/avoidJdkCollectionApiOnFileCollections/groovy",files="build.gradle[tags=bad-zipping-task]"]
+====
+
+<1> *Add file via plus operator*: Plus operator will force runtimeClasspath configuration to be resolved implicitly. Task dependencies will be lost.
```

Comment:
I think this needs a mention of _why_ the plus operator does this.

Reply:
Do you mean the fact that Configuration or FileCollection don't override plus operator?

Reply:
Updated.